### PR TITLE
Conversion between javaapi.data.Value and lf.value.Value

### LINKED ISSUE #15865: Conversion from javaapi.data.Value to lf.value.Value
From https://github.com/digital-asset/daml/issues/13558

Create a new library under language support to implement the conversion between `javaapi.data.Value` and `lf.value.Value` in both directions

### PR DESCRIPTION
**Pull request by chunlokling-da**
fixes #15865 and #15867

### CODE REVIEW

**Review by ray-roestenburg-da — APPROVED** [review]

**Comment by ray-roestenburg-da** [comment]
(the approval means, looks good so far to me @chunlokling-da )

File: language-support/java/json/src/main/scala/com/daml/lf/codegen/json/ValueConversion.scala [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+      )
+    case LfValue.ValueVariant(tycon, variant, value) =>
+      tycon.cata(
+        id => new JData.Variant(fromRefId(id), variant, fromLfValue(value)),
+        new JData.Variant(variant, fromLfValue(value)),
+      )
+
+    case LfValue.ValueContractId(value) => new JData.ContractId(value.coid)
+    case LfValue.ValueList(values) =>
+      JData.DamlList.of(values.toImmArray.toList.map(fromLfValue).asJava)
```
Comment (S11001001) [inline language-support/java/json/src/main/scala/com/daml/lf/codegen/json/ValueConversion.scala:95]:
```suggestion
      JData.DamlList.of(values.toImmArray.toSeq.map(fromLfValue).asJava)
```

File: language-support/java/json/src/main/scala/com/daml/lf/codegen/json/ValueConversion.scala [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    case date: JData.Date =>
+      LfValue.ValueDate(Time.Date.assertFromDaysSinceEpoch(date.getValue.toEpochDay.toInt))
+    case party: JData.Party => LfValue.ValueParty(Ref.Party.assertFromString(party.getValue))
+    case bool: JData.Bool => LfValue.ValueBool(bool.getValue)
+    case _: JData.Unit => LfValue.ValueUnit
+  }
+
+  def fromLfValue(lfV: LfValue): JData.Value = lfV match {
+    case LfValue.ValueRecord(tycon, fields) =>
+      val lfFields = fields.toList.map { case (name, value) =>
```
Comment (S11001001) [inline language-support/java/json/src/main/scala/com/daml/lf/codegen/json/ValueConversion.scala:77]:
```suggestion
      val lfFields = fields.toSeq.map { case (name, value) =>
```

File: language-support/java/json/src/main/scala/com/daml/lf/codegen/json/ValueConversion.scala [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+      JData.DamlOptional.of(value.map(fromLfValue).toJava)
+    case LfValue.ValueTextMap(value) =>
+      JData.DamlTextMap.of(value.toHashMap.view.mapValues(fromLfValue).toMap.asJava)
+    case LfValue.ValueGenMap(entries) =>
+      JData.DamlGenMap.of(
+        entries
+          .map { case (v1, v2) =>
+            fromLfValue(v1) -> fromLfValue(v2)
+          }
+          .toList
```
Comment (S11001001) [inline language-support/java/json/src/main/scala/com/daml/lf/codegen/json/ValueConversion.scala:106]:
```suggestion
          .toSeq
          .view
          .map { case (v1, v2) =>
            fromLfValue(v1) -> fromLfValue(v2)
          }
```

File: language-support/java/json/src/main/scala/com/daml/lf/codegen/json/ValueConversion.scala (hunk 2 above)
Comment (S11001001) [reply language-support/java/json/src/main/scala/com/daml/lf/codegen/json/ValueConversion.scala:77]: Honestly we should probably just deprecate `ImmArray#toList`.

File: language-support/java/json/src/test/scala/com/daml/lf/codegen/json/ValueConversionSpec.scala [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      forAll(variantGen)(testRoundTrip)
+    }
+
+    "do record" in {
+      forAll(recordGen)(testRoundTrip)
+    }
+
+    "do unit" in {
+      testRoundTrip(ValueUnit)
+    }
```
Comment (S11001001) [inline language-support/java/json/src/test/scala/com/daml/lf/codegen/json/ValueConversionSpec.scala:105]: * [ ] add `valueGenMapGen` to `ValueGenerators.nestedGen`
* [ ] add `ValueUnit` to the last line in `ValueGenerators.flatGen`

```suggestion
    "do values" in forAll(valueGen)(testRoundTrip)
```

File: language-support/java/json/src/main/scala/com/daml/lf/codegen/json/ValueConversion.scala [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+      LfValue.ValueOptional(
+        optional.getValue.toScala.map(toLfValue)
+      )
+    case textMap: JData.DamlTextMap =>
+      LfValue.ValueTextMap(
+        SortedLookupList(textMap.toMap(JFunction.identity(), toLfValue).asScala.toMap)
+      )
+    case genMap: JData.DamlGenMap =>
+      LfValue.ValueGenMap(
+        ImmArray.from(genMap.toMap(toLfValue, toLfValue).asScala)
```
[the remark this answers is not in the document]
Comment (chunlokling-da) [reply language-support/java/json/src/main/scala/com/daml/lf/codegen/json/ValueConversion.scala:52]: the orders are still not preserved if using `genMap.stream` right? it is returning a entry set 
`return map.entrySet().stream();`

File: daml-lf/transaction-test-lib/src/main/scala/lf/value/test/ValueGenerators.scala [hunk 6]
```diff
@@ -203,6 +203,7 @@ object ValueGenerators {
     recordGen,
     valueOptionalGen,
     valueMapGen,
+//    valueGenMapGen,
```
Comment (ray-roestenburg-da) [inline daml-lf/transaction-test-lib/src/main/scala/lf/value/test/ValueGenerators.scala:206]: remove commented line

File: daml-lf/transaction-test-lib/src/main/scala/lf/value/test/ValueGenerators.scala [hunk 7]
```diff
@@ -157,6 +157,11 @@ object ValueGenerators {
       .listOf(Gen.zip(valueGen, valueGen))
       .map(list => ValueGenMap(list.to(ImmArray)))
 
+  def valueGenUniqueKeysMapGen: Gen[ValueGenMap] =
+    Gen
+      .listOf(Gen.zip(valueGen, valueGen))
+      .map(list => ValueGenMap(list.distinctBy(_._1).to(ImmArray)))
```
Comment (S11001001) [inline daml-lf/transaction-test-lib/src/main/scala/lf/value/test/ValueGenerators.scala:163]: I would just fix `valueGenMapGen` to do this.

File: language-support/java/json/src/test/scala/com/daml/lf/codegen/json/ValueConversionSpec.scala [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+  implicit val noStringShrink: Shrink[String] = Shrink.shrinkAny[String]
+
+  implicit override val generatorDrivenConfig: PropertyCheckConfiguration =
+    PropertyCheckConfiguration(minSuccessful = 1000)
+
+  "value conversion" should {
+    "do values" in {
+      "do values" in forAll(valueGen)(testRoundTrip)
+    }
```
Comment (S11001001) [inline language-support/java/json/src/test/scala/com/daml/lf/codegen/json/ValueConversionSpec.scala:28]: This test obviates all the other cases, so you can just get rid of them.
```suggestion
    "do values" in forAll(valueGen)(testRoundTrip)
```

**Review by remyhaemmerle-da — COMMENTED** [review]
LGTM

**Pull request merged**